packages/esm-patient-vitals-app/src/vitals-biometrics-form/vitals-biometrics-form.component.tsx

```diff
@@ -87,22 +62,152 @@ const VitalsAndBiometricsForm: React.FC<DefaultWorkspaceProps> = ({
   const patient = usePatient(patientUuid);
   const { currentVisit } = useVisit(patientUuid);
   const { data: conceptUnits, conceptMetadata, conceptRanges, isLoading } = useVitalsConceptMetadata();
-  const [hasInvalidVitals, setHasInvalidVitals] = useState(false);
   const [isSubmitting, setIsSubmitting] = useState(false);
   const [muacColorCode, setMuacColorCode] = useState('');
-  const [showErrorNotification, setShowErrorNotification] = useState(false);
-  const [showErrorMessage, setShowErrorMessage] = useState(false);
+  const [showFormSubmissionErrorNotifications, setShowFormSubmissionErrorNotifications] = useState(false);
+
+  const concepts = useMemo(
+    () => ({
+      midUpperArmCircumference: conceptRanges.get(config.concepts.midUpperArmCircumferenceUuid),
+      diastolicBloodPressure: conceptRanges.get(config.concepts.diastolicBloodPressureUuid),
+      systolicBloodPressure: conceptRanges.get(config.concepts.systolicBloodPressureUuid),
+      oxygenSaturation: conceptRanges.get(config.concepts.oxygenSaturationUuid),
+      respiratoryRate: conceptRanges.get(config.concepts.respiratoryRateUuid),
+      temperature: conceptRanges.get(config.concepts.temperatureUuid),
+      weight: conceptRanges.get(config.concepts.weightUuid),
+      height: conceptRanges.get(config.concepts.heightUuid),
+      pulse: conceptRanges.get(config.concepts.pulseUuid),
+    }),
+    [
+      conceptRanges,
+      config.concepts.diastolicBloodPressureUuid,
+      config.concepts.heightUuid,
+      config.concepts.midUpperArmCircumferenceUuid,
+      config.concepts.oxygenSaturationUuid,
+      config.concepts.pulseUuid,
+      config.concepts.respiratoryRateUuid,
+      config.concepts.systolicBloodPressureUuid,
+      config.concepts.temperatureUuid,
+      config.concepts.weightUuid,
+    ],
+  );
+
+  const decimalRefinement = (key: keyof VitalsBiometricsFormData, schema: z.ZodNumber) => {
+    const decimalCountAllowed = getDecimalCountForField(key);
+    return schema.refine(
+      (value) => {
+        if (value === undefined) {
+          return true;
+        }
+        if (!decimalCountAllowed) {
+          return Number.isInteger(value);
+        }
+
+        const decimalCount = value.toString().split('.')?.[1]?.length ?? 0;
```

Comment:
You could also do this mathematically, though I suppose that won't catch `10.0`, but again, it's weird to require a decimal point in that case.

Reply:
Updated the function to check if the number is an integer or not. If it's an integer (10, 10.0, 10.), it won't reach the below point.